File: ExampleScripts/node/schools.js
#!/usr/bin/env node

// For more info on USERS from our API and additional parameters and options that are available,
// please refer to our official Documentation: http://docs.whetstoneeducation.com/#api-Users

// Import all of the helper functions that we need.
const { http, getAuthHeader, authenticate } = require('./utils');

/**
 * This function will return all schools from your allowed districts in a paginated format.
 * @param accessToken The accessToken will authorize you to communicate with our API.
 */
const getAllSchools = async (accessToken) => {
  const endpoint = `/external/schools`;
  const res = await http.get(endpoint, getAuthHeader(accessToken));
  if (res && res.body) {
    return res.body;
  } else {
    return null;
  }
}

/**
 * This function will return a specific school by ID that you have access to.
 * @param accessToken The accessToken will authorize you to communicate with our API.
 * @param schoolId The ID of the school you want to GET.
 */
const getSchoolById = async (accessToken, schoolId) => {
  const endpoint = `external/schools/${schoolId}`;
  const res = await http.get(endpoint, getAuthHeader(accessToken));
  if (res && res.body) {
    return res.body;
  } else {
    return null;
  }
}

/**
 * Update a School's fields by its ID.
 * There are some specific rules here. Please view the docs for up to date information:
 * http://docs.whetstoneeducation.com/#api-Schools-UpdateSchool
 * @param accessToken The accessToken will authorize you to communicate with our API.
 * @param schoolId The ID of the school you want to GET.
 * @param fieldsToUpdate A JSON object representing the fields you want to update on the school.
 */
const updateSchoolById = async (accessToken, schoolId, fieldsToUpdate) => {
  const endpoint = `external/schools/${schoolId}`;
  const res = await http.put(endpoint, fieldsToUpdate, getAuthHeader(accessToken));
  if (res && res.body) {
    return res.body;
  } else {
    return null;
  }
}

/**
 * This is where we actually run all of our scripts above for testing purposes.
 */
(async () => {
  try {
    // You must first always run the authenticate method to obtain an access token.
    // The Access Token will last for 24 hours--so you can store it and use it multiple times!
    const authToken = await authenticate();
    if (authToken) {

      // Get All Schools
      const schools = await getAllSchools(authToken);
      console.log(schools);

      // Get School By ID
      const school = await getSchoolById(authToken, "60c141a5f3c96200128a2098");
      console.log(school);

      // Update A School By ID
      // This is also an example of updating Observation Groups on Schools
      // There are specific rules surrounding how this works so please refer to the docs as well:
      // http://docs.whetstoneeducation.com/#api-Schools-UpdateSchool
      // In this example we are adding a User with a TEACHER role to the TEACHERS group as an observee.
      const fieldsToUpdate = {
        district: "5d24f84b0ad5dc0019f115e3", // Required
        name: "My Updated School External API",

      // WARNING: When updating OBSERVATION GROUPS--it will REPLACE the entire object in the database.
      // This means if you want to ADD 1--then you need to include all of the other ones that exist in the request.
      // If you do NOT do this--it will add only the ones you send and delete the rest.

        observationGroups: [
          {
            _id: "60fee7e5b6f902001002304b", // Group ID
            name: "Teachers", // Group Name
            observees: [
                "60f57cbeac1e5c001199bb3c" // UserID
            ]
          }
        ]
      };
      const updatedSchool = await updateSchoolById(authToken, "60fee7e5b6f902001002304a", fieldsToUpdate);
      console.log(updatedSchool);
    }
  } catch (err) {
    console.error(err);
  }
})();

File: ExampleScripts/python/schools.py
#!/usr/bin/python

import requests
import pprint
from config import BASE_URL
from utils import authenticate

'''
This function will return all schools from your allowed districts in a paginated format.
'''
def getAllSchools(access_token):
    response = requests.get(
        BASE_URL + '/external/schools',
        headers={'Authorization': 'Bearer ' + access_token}
    )
    if response.status_code == 200:
        response_json = response.json()
        return response_json['data']
    else:
        raise Exception('Failed to get all schools')

'''
This function will return a specific school by its ID.
'''
def getSchoolById(access_token, school_id):
    response = requests.get(
        BASE_URL + '/external/schools/' + school_id,
        headers={'Authorization': 'Bearer ' + access_token}
    )
    if response.status_code == 200:
        response_json = response.json()
        return response_json
    else:
        raise Exception('Failed to get school by ID')

'''
Update a School's fields by its ID.

There are some specific rules here. Please view the docs for up to date information:
http://docs.whetstoneeducation.com/#api-Schools-UpdateSchool
'''
def updateSchoolById(access_token, school_id, fields_to_update):
    response = requests.put(
        BASE_URL + '/external/schools/' + school_id,
        headers={'Authorization': 'Bearer ' + access_token},
        json=fields_to_update
    )
    if response.status_code == 200:
        response_json = response.json()
        return response_json
    else:
        raise Exception('Failed to update school by ID')

# Authenticate
token = authenticate()

# Get All Schools
schools = getAllSchools(token)
pprint.pprint(schools)

# Get School By ID
school_id = '59401f146353df000c415340'
school = getSchoolById(token, school_id)
pprint.pprint(school)

# Update School By ID
# There are some specific rules here. Please view the docs for up to date information:
# http://docs.whetstoneeducation.com/#api-Schools-UpdateSchool
fields_to_update = {
    'district': '5ba264d5cd5f35424ee05d54',
    'name': 'My Updated School External API',

    # WARNING: When updating OBSERVATION GROUPS--it will REPLACE the entire object in the database.
    # This means if you want to ADD 1--then you need to include all of the other ones that exist in the request.
    # If you do NOT do this--it will add only the ones you send and delete the rest.

    'observationGroups': [
        {
            '_id': '5d717a856140150012191e16',
            'name': 'Teachers',
            'observees': [
                '5a7c8bed98a23600136c8db8'
            ]
        }
    ]
}
pprint.pprint(fields_to_update)
updatedSchool = updateSchoolById(token, school_id, fields_to_update)
pprint.pprint(updatedSchool)



